lib/kdl/include/kdl/string_format.h

```diff
@@ -0,0 +1,301 @@
+        std::stringstream result;
+        result << prefix << str_plural(count, singular, plural) << suffix;
+        return result.str();
+    }
+
+    /**
+     * Trims the longest prefix and the longest suffix consisting only of the given characters from the given string.
+     *
+     * @param s the string to trim
+     * @param chars the characters that should be trimmed from the start and from the end of the given string
+     * @return the trimmed string
+     */
+    inline std::string str_trim(const std::string_view& s, const std::string_view& chars = Whitespace) {
+        if (s.empty() || chars.empty()) {
+            return std::string(s);
+        }
+
+        const auto first = s.find_first_not_of(chars);
+        if (first == std::string::npos) {
+            return "";
+        }
+
+        const auto last = s.find_last_not_of(chars);
+        if (first > last) {
+            return "";
+        }
+
+        return std::string(s.substr(first, last - first + 1));
+    }
+
+    /**
+     * Convers the given string to lowercase using the C locale.
+     *
+     * @param str the string to convert
+     * @return the converted string
+     */
+    inline std::string str_to_lower(const std::string_view& str) {
+        auto result = std::string();
+        result.reserve(str.size());
+
+        for (const auto c : str) {
+            result.push_back(static_cast<std::string::value_type>(std::tolower(c)));
+        }
+
+        return result;
+    }
+
+    /**
+     * Convers the given string to uppercase using the C locale.
+     *
+     * @param str the string to convert
+     * @return the converted string
+     */
+    inline std::string str_to_upper(const std::string_view& str) {
+        auto result = std::string();
+        result.reserve(str.size());
+
+        for (const auto c : str) {
+            result.push_back(static_cast<std::string::value_type>(std::toupper(c)));
+        }
+
+        return result;
+    }
+
+    /**
+     * Converts the first character and any character following one (or multiple) of the given delimiters to upper case.
+     *
+     * For example, calling capitalize("by the power of greyscull!", " ") would result in the string "By The Power Of
+     * Greyscull!"
+     *
+     * @param str the string to capitalize
+     * @param delims the delimiters, defaults to whitespace
+     * @return the capitalized string
+     */
+    inline std::string str_capitalize(const std::string_view& str, const std::string_view& delims = Whitespace) {
+        auto result = std::string();
+        result.reserve(str.size());
+
+        bool initial = true;
+        for (const auto c : str) {
+            if (delims.find(c) != std::string::npos) {
+                initial = true;
+                result.push_back(c);
+            } else if (initial) {
+                result.push_back(static_cast<std::string::value_type>(std::toupper(c)));
```

Comment:
The strategy we're taking is allowing strings to be ASCII or UTF-8, but only implementing case-insensitivity for ascii a-z. this is reasonable for TrenchBroom, I think. If we want full unicode case sensitivity we'd either need to use Qt (slow because you'd have to construct a QString for every case-insensitive operation) or pull in some 3rd party library that implements it for utf-8.

I'd suggest documeting this explicitly and implementing our own  `char kdl::tolower(char c)` and toupper so we have less stuff depending on the locale being set to "C". (other kdl string code probably still requries the C locale to work, but reducing the amount that assumes this is probably good, especially since tolower/toupper are so easy to implement for ascii).

Reply:
done